Source/Core/Resource.js

```diff
@@ -0,0 +1,534 @@
+     * @param {String} [options.responseType] The type of response.
+     * @param {String} [options.method='GET'] The method to use.
+     * @param {Object} [options.data] Data that is sent with the resource request.
+     * @param {String} [options.overrideMimeType] Overrides the MIME type returned by the server.
+     * @param {DefaultProxy} [options.proxy] A proxy to be used when loading the resource.
+     * @param {Boolean} [options.isDirectory=false] The url should be a directory, so make sure there is a trailing slash.
+     * @param {Function} [options.retryCallback] The Function to call when a request for this resource fails. If it returns true, the request will be retried.
+     * @param {Number} [options.retryAttempts=0] The number of times the retryCallback should be called before giving up.
+     * @param {Request} [options.request] A Request object that will be used. Intended for internal use only.
+     *
+     * @constructor
+     */
+    function Resource(options) {
+        options = defaultValue(options, defaultValue.EMPTY_OBJECT);
+
+        //>>includeStart('debug', pragmas.debug);
+        Check.typeOf.string('options.url', options.url);
+        //>>includeEnd('debug');
+
+        this._url = '';
+        this._templateValues = encodeValues(defaultValue(options.templateValues, {}));
+        this._queryParameters = defaultValue(options.queryParameters, {});
+        this._isDirectory = defaultValue(options.isDirectory, false);
+        this.url = options.url;
+
+        /**
+         * Additional HTTP headers that will be sent with the request.
+         *
+         * @type {Object}
+         */
+        this.headers = defined(options.headers) ? clone(options.headers) : {};
+
+        /**
+         * A Request object that will be used. Intended for internal use only.
+         *
+         * @type {Request}
+         */
+        this.request = options.request;
+
+        /**
+         * The type of response expected from the request.
+         *
+         * @type {String}
+         */
+        this.responseType = options.responseType;
+
+        /**
+         * The method to use for the request.
+         *
+         * @type {String}
+         */
+        this.method = defaultValue(options.method, 'GET');
+
+        /**
+         * Data to be sent with the request.
+         *
+         * @type {Object}
+         */
+        this.data = options.data;
+
+        /**
+         * Overrides the MIME type returned by the server.
+         *
+         * @type {String}
+         */
+        this.overrideMimeType = options.overrideMimeType;
+
+        /**
+         * A proxy to be used when loading the resource.
+         *
+         * @type {DefaultProxy}
+         */
+        this.proxy = options.proxy;
+
+        /**
+         * Function to call when a request for this resource fails. If it returns true, the request will be retried.
+         *
+         * @type {Function}
```

Comment:
You can define the actual callback signature in JSDoc, look in CallbackProperty.js for an example.